Gravity-Bridge/Gravity-Bridge
# ICA upgrade prep

Pull request contains:
* ICA implementation
* ICA tests
* Upgrade handler

Notes:
* This pull request is re-created #193
* ICAPlanName set to "ica" `might be changed`
* Upgrade handler name set to "icav3" `might be changed`
* Added ICAAuth module ( regular implementation) (ibc-go v3)
* Implemented both ICA Host & Controller (ibc-go v3)
* Added third_party/proto/ibc
* ibc-go `v3.1.0` -> `v3.2.0`
* sdk `v0.45.6` -> `v0.45.7`

Integrated tests:
* Added new test_type - `ICA_TEST`
* hermes for tests changed `"0.13.0 -> 0.15.0"`
* gaiad binary replaced with forked (ICA auth included) , version 7.0.0 `might be replaced with icad after migration to ibc-go v4`
* Added connection creation to test_runner bootstrap, changed channel creation (hermes changed), added delay after `create connection`, 
* Added new test mod `ica.rs`
* Updated rs proto. (Done manually, due to a lot of changes by proto_build)

Test workflow:
* Create connection between chains
* Get connection IDs on both chains
* Create interchain account on both chains
* Send tokens to interchain accounts on both chains from regular addrs
* Delegate tokens to validator from interchain account. Send delegation TX from counter-party chain by SubmitTX
* Check delegation shares


Other notes:
* Migration to v4 possible after bech32 module upgrade ( due to v3 relations)
* Allowed host messages might be filled in upgrade handler: e.g.  ibc, bank, stake, etc. allow all = '*'  , also messages can be add/changed via parameter change proposal. 
* Allowed messages for tests workflow added to `InitChainer`

All interchain interactions proceed via `SubmitTx` .
`SubmitTx` contain `native for counter-party chain` messages, like `/cosmos.staking.v1beta1.MsgDelegate` , so tests include only by 1 `SubmitTx` for both side
